goto_symex: rename type symbols only when needed [blocks: #3652]

kroening opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

kroening committed:
goto_symex: rename type symbols only when needed

Renaming a type is rarely needed: this is necessary only if the type
contains a struct that contains a variable-sized member, which is a gcc
extension.  The key benefit of not renaming in all other cases is that
struct and union tag types no longer get expanded.

kroening commented:
This fails on jbmc regressions -- apparently types created by the Java object factory are sometimes expanded to ID_struct.

tautschnig commented:
> This fails on jbmc regressions -- apparently types created by the Java object factory are sometimes expanded to ID_struct.

This might as well be something that the simplifier does. See some of the commits in #2064.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -625,6 +625,52 @@ void goto_symex_statet::rename_address(
+           !array_type.size().is_constant();
+  }
+  else if(
+    type.id() == ID_struct || type.id() == ID_union || type.id() == ID_class)
```
I don't think `type.id() == ID_class` should ever evaluate to true. A `class` is represented as `ID_struct` with `ID_C_class` set to `true`. If I'm wrong in saying this then we've got a lot more work to do elsewhere.

kroening replied:
Yes, these should be cleaned out.

kroening replied:
#3635

kroening commented:
Note #3634 -- the assertion there catches this problem.

tautschnig commented:
I think it would be good to factor out the individual commits into PRs of their own. They need proper justifications, but seem to be useful in their own right. And also at least the first one also appears in another PR.

kroening commented:
The bugfixes are now in separate PRs.
For this to work in jbmc, the problem in the Java object factory needs to be fixed first, as part of #3634.

tautschnig commented:
@kroening Are all the depended-on changes now PR'ed, or is there more work necessary?

tautschnig committed:
Make tests oblivious to tag type expansion

The regular expressions now accept both the tag names as well as the expanded
types/class names.

tautschnig committed:
Accept base type equality

Until we have resolved partial tag expansion everywhere we need to use
base_type_eq for cases where both tags and expanded types may occur.

tautschnig committed:
has_subtype: handle struct/union tag types

Just like symbol or enum tags, we need to look into the expanded tags.

tautschnig committed:
String refinement: handle both struct and struct tag types

struct tags need to be expanded.

tautschnig reviewed (approved):
Approving, but this PR now includes commits of my own so @kroening (and other reviewers as other code is now being touched) should please comment as well.

smowton commented:
Any chance of unifying the make-change and check-if-change-needed paths? For example, by passing a `const typet &` and a `typet *` in parallel, which is null in the check-if-needed case and non-null otherwise?

tautschnig commented:
> Any chance of unifying the make-change and check-if-change-needed paths?

Good point, but I'm a bit hesitant here: not doing the renaming also spares us putting anything in `l1_types`, which we'd otherwise likely want to be doing lookups in. Doable, but I'd for now prefer to keep this one simple until profiling tells us that it is one of the bottlenecks worth tackling.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 74d15488dee69ec0f449a4fcd2c2ff4cdca19dde).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98362464

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -618,12 +618,63 @@ void goto_symex_statet::rename_address(
+    {
+      // be careful, or it might get cyclic
+      if(component.type().id() != ID_pointer)
+        return requires_renaming(component.type(), ns);
```
This only looks at the first non-pointer component of the struct, which is strange, but I cannot say whether this is correct because the function is not documented, so please document.

smowton replied:
Oooh, I missed that, @kroening pretty sure this is a real bug

smowton replied:
(suggests a new test -- AFAIK GCC allows such variable-size structs?)

tautschnig replied:
I think I've found a way to construct a failing test, I'll post a follow-up PR to include a test, a fix, and the suggested documentation.

tautschnig replied:
The failure I thought I had found was one caused by integer overflow - I haven't actually managed to produce a proper failure. See #3918 for my attempt and the cleanup.

kroening commented:
The base_type_eq shouldn't be needed; however, this is a step in the right direction.

tautschnig commented:
> The base_type_eq shouldn't be needed; [...]

Yes, it _shouldn't_ be, but thus far it strictly is needed for Java input programs (as regression tests confirm).
